HomeWork Java Collections, Stream-API v1

AKnazzz opened the pull request:
Выполнение домашней работы в ветке dev. 

**ADD:** 
- [x] - Решены задачи Task 1-9;
- [x] - Проверено прохождение всех текущих Unit тестов;
- [x] - Добавлены комментарии.

AKnazzz committed:
ADD: Task1 solution. Все тесты прошли.

AKnazzz committed:
ADD: Tasks 2-7 solutions. Все тесты прошли.
EDIT: Task 1 codestyle.

AKnazzz committed:
ADD: Tasks 8-9 solutions. В 9 задаче добавлены комментарии.

genroelgvozo commented on src/main/java/tasks/Task1.java:
```diff
@@ -23,6 +25,20 @@ public Task1(PersonService personService) {
+    return personIds.stream() // Сортируем список по порядку из personIds
+            .map(personsMap::get)
+            .filter(Objects::nonNull) // можно убрать, если гарантируется что все Id ---> Person
+            .collect(Collectors.toList());
```
ну к коду комментарии не нужны (разве что в целях дз - но мне вроде и так все понятно зачем тут))

А так все гуд

genroelgvozo replied:
разве что можно сразу в мапу, сет промежуточный тебе не нужен

AKnazzz replied:
Учтено

genroelgvozo commented on src/main/java/tasks/Task1.java:
```diff
@@ -23,6 +25,20 @@ public Task1(PersonService personService) {
+Оценка асимптотики:
+---> Создание Map: O(n), где n — количество элементов в Set<Person>;
+---> Итерация по personIds: O(m), где m — кол-во Id.
+===> Асимптотика работы метода findOrderedPersons будет: == O(n + m).
```
если уж строго - то O(n), где n кол-во айдишек, ибо кол-во персон ими ограничено)

genroelgvozo commented on src/main/java/tasks/Task2.java:
```diff
@@ -1,21 +1,30 @@
                                                      int limit) {
-    return new ArrayList<>();
+    if (persons1 == null || persons2 == null || limit < 0) {
+      throw new IllegalArgumentException("Коллекции Person не должны быть null, а величина limit должна быть неотрицательна.");
```
Я о таком не говорил, но в целом настолько "естественные" ограничения не проверяются обычно - код засоряется, а смысла ноль. Странно если кто-то захочет combine коллекции и null сделать. Или взять -2 элемента) это он сам виноват.
Поэтому такие проверки обычно не делают, тут очевидно на уровне контракта метода что передавать, а разницы между NPE и IAE большой не будет для клиента если ошибется)

genroelgvozo replied:
Касается всех таких проверках в других задачах

AKnazzz replied:
Убрано во всем коде!

genroelgvozo commented on src/main/java/tasks/Task5.java:
```diff
@@ -23,6 +23,22 @@ public Task5(PersonConverter personConverter) {
+              if (areaId != null) {
+                dto.setAreaId(areaId);
+              }
+              return dto;
```
а почему ты доконверчиваешь сам? у personConverter есть второй метод)

AKnazzz replied:
Да увидел второй метод --- с ним код ещё проще стал

genroelgvozo commented on src/main/java/tasks/Task6.java:
```diff
@@ -19,6 +20,26 @@ public class Task6 {
+      Integer personId = person.id();
+      Set<Integer> areaIds = personAreaIds.get(personId);
+
+      if (areaIds != null) {
```
ну очень странная проверка - специально кто-то положил personId а значение null)
Это все решается контрактом - ты ожидаешь не налл, виноват тот кто выше)

AKnazzz replied:
Исправил метод теперь чисто

genroelgvozo commented on src/main/java/tasks/Task6.java:
```diff
@@ -19,6 +20,26 @@ public class Task6 {
+        for (Integer areaId : areaIds) {
+          String areaName = areaIdToNameMap.get(areaId);
+          if (areaName != null) {
+            descriptions.add(personName + " - " + areaName);
```
В целом скорость работы ок
Куча лишних проверок, и вложенности из-за этого - не ок

В идеале как бы надо
1. Вынести склейку в метод, аля "формат". 
2. Метод бы принимает лучше персону и арейку целиком - пускай формат сам решает какие поля ему нужны, так проще его поправить, без правки алгоритма (а сейчас у тебя только имя арейки в мапе, а может появиться что-то еще)
3. Ну и в целом все это легко через флатмап пишется, и даже ифы если уж очень хочется пишутся через filter не добавляя такой вложенности

AKnazzz replied:
Добавил метод для форматирования строки + изменил структуру метода, избавился от лишних вложений

genroelgvozo commented on src/main/java/tasks/Task9.java:
```diff
@@ -21,69 +15,83 @@
+   */
   public List<String> getNames(List<Person> persons) {
-    if (persons.size() == 0) {
+    if (persons == null || persons.isEmpty()) {   // изменено для улучшения читаемости
```
а оно теперь надо?

genroelgvozo commented on src/main/java/tasks/Task9.java:
```diff
@@ -21,69 +15,83 @@
+  /*
+  Удаление первого элемента с помощью persons.remove(0) изменяет исходный список, что может быть нежелательно.
+  Проверка на пустоту списка выполняется, но удаление элемента все равно происходит, если список не пуст.
+  Удаление элемента из начала списка неэффективно для больших списков.
```
LinkedList - вполне эффективно я скажу)
А в чем еще проблема то?

AKnazzz replied:
Вместо удаления элемента мы пропускаем элемент стриме. Это сохраняет исходный список неизменным и более эффективно.
Удалил проверку на пустой список - тк если на входе пустой список по стрим вернёт пустой список + в данной проверке возвращается иммутабельный список,   а я предполагаю что с ним продолжится работа

genroelgvozo replied:
Последний точно забудь аргумент, продолжится работа - да, надо ли возвращать мутабельным - нет
Его возможно сконвертят или сериализуют, но точно не будут модифицировать исходный который ты вернул, это антипаттерн

genroelgvozo commented on src/main/java/tasks/Task9.java:
```diff
@@ -21,69 +15,83 @@
+    }
+    return getNames(persons).stream()
+            .filter(name -> name != null && !name.isEmpty()) // Фильтруем null и пустые строки
+            .collect(Collectors.toSet()); // Собираем в Set
```
ну это ты уже отсебятины какой-то - исходный код фильтровал? Все таки я просил поправить "качество" кода, а не менять логику)

AKnazzz replied:
Отсебятину убрал -> метод исправлен

genroelgvozo commented on src/main/java/tasks/Task9.java:
```diff
@@ -21,69 +15,83 @@
+            Optional.ofNullable(person.secondName()).orElse(""),
+            Optional.ofNullable(person.firstName()).orElse(""),
+            Optional.ofNullable(person.middleName()).orElse("") // ФИО
+    ).trim();
```
"Негрозов  Александрович" - два пробела

genroelgvozo commented on src/main/java/tasks/Task9.java:
```diff
@@ -21,69 +15,83 @@
     }
-    return map;
+    return persons.stream()
+            .distinct() // если нужно убрать дубликаты по объектам Person
```
"если нужно" - а нужно?)
а в мапе могут быть дубликаты?
а точно это ТАК надо делать?)

AKnazzz replied:
Разобрался с методом = eсли придёт совпадающий id персоны код упадёт -> добавляем значение только в том случае, если в мапе ключа еще нет

genroelgvozo commented on src/main/java/tasks/Task9.java:
```diff
@@ -21,69 +15,83 @@
-    return has;
+    Set<Person> set1 = new HashSet<>(persons1); // Создаем HashSet для первой коллекции
+    set1.retainAll(persons2); // Оставляем только те элементы, которые есть в persons2
+    return !set1.isEmpty(); // Если остались элементы, значит есть совпадения
```
работать будет, но так же медленно (если посмотришь как написан retainAll)

AKnazzz replied:
Переделал на anyMatch - он должен быть быстрее

genroelgvozo replied:
а с ретейном то можно так же быстро?) anyMatch тут вообще не причем))

genroelgvozo commented on src/main/java/tasks/Task9.java:
```diff
@@ -94,5 +102,16 @@ void listVsSet() {
+    4. Когда мы вызываем toString() на множестве, оно возвращает строку с элементами, отсортированными по возрастанию: [1, 2, 3, ..., 10000].
+
+    Таким образом, хотя порядок элементов в snapshot и set отличается, их содержимое одинаково.
+    Поэтому вызов метода toString() на обоих объектах вернет одну и ту же строку: все числа от 1 до 10000 в отсортированном виде.
```
Написано много, но куда-то суть потерял
Пункт 4 - а почему?)
это же множество, ты пишешь порядок не имеет значения, а в п.4. вдруг отсортированные))
Все остальные пункты как будто бы тривиальны, это просто то что происходит в методе

AKnazzz replied:
Добавил коммент в коде:  В множестве set все элементы хранятся по бакетам, номер бакета определяется хэш функцией для каждого элемента . Но для целых чисел хэш-код является самим числом - поэтому они будут располагаться по возрастанию в самом HashSet независимо от порядка в котором добавлялись. В итоге toString() пойдёт по бакетам и даст одинаковый результат как для упорядоченной коллекции так и для множества, т.к. они с целыми числами.

genroelgvozo replied:
Так, беру пустую мапу, и добавлю 1 2 16 33 - вывожу, а они не по порядку. Вроде все что ты объяснил подходит и под этот случай, почему не по возрастанию?

AKnazzz committed:
FIX: Task1-Task9 внесены правки:
- везде убраны излишние проверки на null;
- доработаны Задача 1,5,6 и 9 - согласно замечаниям.

genroelgvozo commented on src/main/java/tasks/Task9.java:
```diff
@@ -15,83 +24,56 @@
-    Создаем Map, где ключом является id человека, а значением — строковое представление ФИО Person.
-   */
+  // Изначально map создаётся размером 1 - лучше сделать без указания начального размера
+  // Если придёт совпадающий id персоны код упадёт -> добавляем значение только в том случае, если в мапе ключа еще нет
```
3 задачи до этого собирал в мапу через стрим, а тут чего вдруг забыл как?)
